geoxarray/accessor.py

```diff
@@ -50,53 +73,336 @@ class _SharedGeoAccessor(object):
+                self._dim_map[self._vertical_dim] = "vertical"
+            if self._time_dim is not None:
+                self._dim_map[self._time_dim] = "time"
+
+        return self._dim_map
 
-    @cf_grid_mapping_parameters.setter
-    def cf_grid_mapping_parameters(self, value):
-        self._cf_grid_mapping_parameters = value
-        # reset the CRS object so it gets recalculated
-        self._crs = None
+    @property
+    def dims(self):
+        """Get preferred dimension names in order."""
+        return tuple(self.dim_map.get(dname, dname) for dname in self._obj.dims)
+
+    @property
+    def sizes(self):
+        """Get size map with preferred dimension names."""
+        # return the same type of object as xarray
+        sizes_dict = {}
+        for dname, size in self._obj.sizes.items():
+            sizes_dict[self.dim_map.get(dname, dname)] = size
+        return self._obj.sizes.__class__(sizes_dict)
+
+    def write_dims(self, inplace: bool = False):
+        """Rename object's dimensions to match geoxarray's preferred dimension names."""
+        obj_copy = self._get_obj(inplace)
+        return obj_copy.rename(self.dim_map)
+
+
+@xr.register_dataset_accessor("geo")
+class GeoDatasetAccessor(_SharedGeoAccessor):
+    """Provide Dataset geolocation helper functions from a `.geo` accessor."""
+
+    def set_dims(
+        self,
+        x: Optional[str] = None,
+        y: Optional[str] = None,
+        vertical: Optional[str] = None,
+        time: Optional[str] = None,
+        inplace: bool = False,
+    ):
+        """Tell geoxarray the names of the provided dimensions in this Dataset.
+
+        Parameters
+        ----------
+        x
+        y
+        vertical
+        time
+        inplace
+
+        """
+        all_dims = {
+            "x": x,
+            "y": y,
+            "vertical": vertical,
+            "time": time,
+        }
+
+        obj_copy = self._get_obj(inplace)
+        # tell the dim_map property to produce the "as-is" dim map
+        obj_copy.geo._dim_map = None
+        dim_map = obj_copy.geo.dim_map
+        for data_arr in self._obj.data_vars.values():
+            dims = {k: v for k, v in all_dims.items() if v in data_arr.dims}
+            if not dims:
+                continue
+            var_dim_map = data_arr.geo.set_dims(**dims, inplace=True).geo.dim_map
+            self._update_gx_dim_dict(dim_map, var_dim_map)
+
+        # update our attributes
+        for dim_name, gx_dim_name in dim_map.items():
+            if gx_dim_name is None:
+                continue
+            setattr(obj_copy.geo, f"_{gx_dim_name}_dim", dim_name)
+        # tell the dim_map to get regenerated
+        obj_copy.geo._dim_map = None
+        return obj_copy
+
+    def _update_gx_dim_dict(self, old, new):
+        for k, v in new.items():
+            if v in ("x", "y", "time", "vertical"):
+                old[k] = v
+        return old
+
+    def set_crs(self, crs=None, variables=None):
```

Comment:
I have hidden `set_crs` on `rioxarray` from users as it lost the `CRS` so easily. I instead recommend `write_crs` as it makes the CRS stick around better.